msprime/intervals.py

```diff
@@ -168,134 +160,175 @@ def sequence_length(self):
 
         # Need to check regions before & after sliced region are filled out:
-        if start != 0:
-            if rate[0] == 0:
+        if left != 0:
+            if np.isnan(rate[0]):
                 position[0] = 0  # Extend
             else:
-                rate = np.insert(rate, 0, 0)  # Prepend
+                rate = np.insert(rate, 0, np.nan)  # Prepend
                 position = np.insert(position, 0, 0)
-        if end != self.position[-1]:
-            if rate[-1] == 0:
+        if right != self.position[-1]:
+            if np.isnan(rate[-1]):
                 position[-1] = self.sequence_length  # Extend
             else:
-                rate = np.append(rate, 0)  # Append
+                rate = np.append(rate, np.nan)  # Append
                 position = np.append(position, self.position[-1])
-        return self.__class__(position, rate, start_position=start, end_position=end)
+        return RateMap(position=position, rate=rate)
+
+    def unknown_intervals(self):
+        num_unknown = np.sum(self._unknown)
+        out = np.empty((num_unknown, 2))
+        out[:, 0] = self.left[self._unknown]
+        out[:, 1] = self.right[self._unknown]
+        return out
+
+    def __iter__(self):
+        # The clinching argument for using mid here is that if we used
+        # left instead we would have
+        #   RateMap([0, 1], [0.1]) == RateMap([0, 100], [0.1])
+        # by the inherited definition of equality since the dictionary items
+        # would be equal.
+        # Similarly, we only return the midpoints of known intervals
+        # because NaN values are not equal, and we would need to do
+        # something to work around this. It seems reasonable that
+        # this high-level operation returns the *known* values only
+        # anyway.
+        yield from self.mid[self._known]
+
+    def __len__(self):
+        return np.sum(self._known)
 
     def __getitem__(self, key):
-        """
-        Use slice syntax for obtaining a rate map subset. E.g.
-            >>> rate_map_4m_to_5m = rate_map[4e6:5e6]
-        """
-        if not isinstance(key, slice) or key.step is not None:
-            raise TypeError("Only interval slicing is supported")
-        start, end = key.start, key.stop
-        if start is not None and start < 0:
-            start += self.sequence_length
-        if end is not None and end < 0:
-            end += self.sequence_length
-        return self.slice(start=start, end=end, trim=True)
+        if isinstance(key, slice):
+            if key.step is not None:
+                raise TypeError("Only interval slicing is supported")
+            return self.slice(key.start, key.stop)
+        if isinstance(key, numbers.Number):
+            index = self.find_index(key)
+            return self.rate[index]
+        raise KeyError("Key {key} not in map")
+
+    def __str__(self):
+        data = [
+            [
+                [f"{left:.10g}"],
```

Comment:
I assume this just prints out the unknown regions as "NaN", which is nice?

Reply:
Yep.